Add rowset id generator to FE and BE

yiguolei opened the pull request.

imay commented on be/src/runtime/fe_based_rowset_id_generator.cpp:
```diff
@@ -0,0 +1,108 @@
+}
+
+OLAPStatus FeBasedRowsetIdGenerator::get_next_id(RowsetId* gen_rowset_id) {
+    std::lock_guard<std::mutex> l(_lock);
```
std::mutex is too heavy for get next id. In most cases, it is a only add operation

yiguolei replied:
Done

imay commented on fe/src/main/java/org/apache/doris/catalog/RowsetIdGenerator.java:
```diff
@@ -0,0 +1,71 @@
+
+import com.google.common.collect.Maps;
+
+public class RowsetIdGenerator {
```
Can we name it a PersistedIdGernator? I think it can not only be used for RowsetId, but also be used for other id

yiguolei replied:
Maybe it could not. because rowsetid generator could have many implementations, for example, we could scan all rowset id in storage engine at start time and then generate rowset it one by one. In this case it is not persistent.

imay replied:
For Id generator logic, we can use a class.
Then you can let RowsetIdGenerator contain it.

imay commented on gensrc/thrift/FrontendService.thrift:
```diff
@@ -604,6 +604,12 @@ struct TSnapshotLoaderReportRequest {
+// (start_rowset_id, end_rowset_id) not include start and end
+struct TGetRowsetIdResponse {
+    1: optional i64 start_rowset_id = -1
+    2: optional i64 end_rowset_id = -1
```
I prefer [start, end) with start_id and id_count

yiguolei replied:
done

yiguolei commented on be/src/runtime/fe_based_rowset_id_generator.cpp:
```diff
@@ -0,0 +1,108 @@
+#include "gen_cpp/HeartbeatService_types.h"
+#include "runtime/client_cache.h"
+
+#include "runtime/fe_based_rowset_id_generator.h"
```
Done

yiguolei commented on be/src/runtime/fe_based_rowset_id_generator.cpp:
```diff
@@ -0,0 +1,108 @@
+    // Call fe to get latest user resource
+    Status master_status;
+    // using 500ms as default timeout value    
+    FrontendServiceConnection client(_exec_env->frontend_client_cache(),
```
done

yiguolei committed:
Add local unique rowset id generator

Fix bug

Compile success

Build unit test successfully

Rowsetid version error

Add rowsetid gc logic

Fix bug

Add unique rowset id unit test

gaodayue commented:
Can anyone explain what's the purpose of this PR and give a brief description of what's changed?

yiguolei committed:
Using low as rowset version

yiguolei committed:
Set rowset id to 0 when convert from old tablet

kangpinghuang commented on be/src/agent/agent_server.cpp:
```diff
@@ -426,7 +426,11 @@ void AgentServer::make_snapshot(TAgentResult& return_value,
     vector<string> error_msgs;
     TStatusCode::type status_code = TStatusCode::OK;
-    return_value.__set_snapshot_version(PREFERRED_SNAPSHOT_VERSION);
+    int32_t return_snapshot_version = PREFERRED_SNAPSHOT_VERSION;
```
add some comment here to describe the version meaning.

yiguolei replied:
Done

kangpinghuang commented on be/src/olap/storage_engine.h:
```diff
@@ -196,6 +197,12 @@ class StorageEngine {
+
+    bool rowset_id_in_use(RowsetId& rowset_id) { return _rowset_id_generator->id_in_use(rowset_id); };
+
+    void release_rowset_id(RowsetId& rowset_id) { return _rowset_id_generator->release_id(rowset_id); };
```
add some comments here.

yiguolei replied:
Done

kangpinghuang commented on be/src/olap/rowset/rowset_id_generator.h:
```diff
@@ -28,27 +28,17 @@ class OlapMeta;
-    std::mutex _lock;
-    RowsetId _next_id = -1;
-    RowsetId _id_batch_end = -1;
+    virtual void release_id(RowsetId& rowset_id) = 0;
```
add comment to describe the api

yiguolei replied:
Done

kangpinghuang commented on be/src/olap/rowset/unique_rowset_id_generator.h:
```diff
@@ -0,0 +1,48 @@
+    // generator a id according to data dir
+    // rowsetid is not globally unique, it is dir level
+    // it saves the batch end id into meta env
+    OLAPStatus next_id(RowsetId* rowset_id); 
```
```suggestion
    OLAPStatus next_id(RowsetId* rowset_id) override;
```

yiguolei replied:
Done

kangpinghuang commented on be/src/olap/rowset/unique_rowset_id_generator.h:
```diff
@@ -0,0 +1,48 @@
+    // it saves the batch end id into meta env
+    OLAPStatus next_id(RowsetId* rowset_id); 
+
+    bool id_in_use(RowsetId& rowset_id);
```
```suggestion
    bool id_in_use(const RowsetId& rowset_id) override;
```

yiguolei replied:
Done

kangpinghuang commented on be/src/olap/rowset/unique_rowset_id_generator.h:
```diff
@@ -0,0 +1,48 @@
+
+    bool id_in_use(RowsetId& rowset_id);
+
+    void release_id(RowsetId& rowset_id);
```
```suggestion
    void release_id(const RowsetId& rowset_id) override;
```

yiguolei replied:
Done

kangpinghuang commented on be/src/olap/olap_common.h:
```diff
@@ -241,7 +246,83 @@ typedef std::set<uint32_t> UniqueIdSet;
+        }
+    }
+
+    friend std::ostream &operator<<(std::ostream &out, const RowsetId& rowset_id) {
```
```suggestion
    friend std::ostream& operator<<(std::ostream &out, const RowsetId& rowset_id) {
```

yiguolei replied:
Done

kangpinghuang commented on be/src/olap/olap_common.h:
```diff
@@ -241,7 +246,83 @@ typedef std::set<uint32_t> UniqueIdSet;
+    int64_t lo = 0;
+
+    void init(const std::string& rowset_id_str) {
+        // for new rowsetid its a 48 hex string
```
```suggestion
        // for new rowsetid its a 48 bytes hex string
```

kangpinghuang commented on be/src/olap/olap_common.h:
```diff
@@ -241,7 +246,83 @@ typedef std::set<uint32_t> UniqueIdSet;
+    void init(int64_t id_version, int64_t high, int64_t middle, int64_t low) {
+        version = id_version;
+        if (low >= MAX_ROWSET_ID) {
+            LOG(FATAL) << "low is too large" << low;
```
```suggestion
            LOG(FATAL) << "low is too large: " << low;
```

yiguolei replied:
Done

kangpinghuang commented on be/src/olap/olap_common.h:
```diff
@@ -241,7 +246,83 @@ typedef std::set<uint32_t> UniqueIdSet;
+            int64_t low = 0;
+            from_hex(&high, rowset_id_str.substr(0, 16));
+            from_hex(&middle, rowset_id_str.substr(16, 16));
+            from_hex(&low, rowset_id_str.substr(32, 16));
```
```suggestion
             from_hex(&_hi, rowset_id_str.substr(0, 16));
             from_hex(&_mi, rowset_id_str.substr(16, 16));
             from_hex(&_lo, rowset_id_str.substr(32, 16));
```

kangpinghuang commented on be/src/olap/olap_common.h:
```diff
@@ -241,7 +246,83 @@ typedef std::set<uint32_t> UniqueIdSet;
+    int8_t version = 0;
+    int64_t hi = 0;
+    int64_t mi = 0;
+    int64_t lo = 0;
```
uint64_t?

yiguolei replied:
UniqueId generator using int64_t so that I use int64_t here

kangpinghuang commented on be/src/olap/olap_common.h:
```diff
@@ -241,7 +246,83 @@ typedef std::set<uint32_t> UniqueIdSet;
+// 8 bit, id version
+// 120 bit, backend random number generated by uuid
+// 64 bit, inc number from 0
+struct RowsetId {
```
Can RowsetId support ++ operator?

kangpinghuang replied:
Can directly use protobuf to do serialize and deserialize?

yiguolei replied:
It is useless using ++ , because it will need to get a rowsetidgenerator object. 
Not use pb as a serialize method, we use string.

kangpinghuang commented on be/src/olap/olap_common.h:
```diff
@@ -241,7 +246,83 @@ typedef std::set<uint32_t> UniqueIdSet;
+        lo = (id_version << 56) + (low & LOW_56_BITS);
+    }
+
+    std::string to_string() const {
```
I think you need to_string and serialize and deserialize api

yiguolei replied:
to_string is ok, because it is a string value in pb.

kangpinghuang commented on be/src/olap/rowset/beta_rowset.cpp:
```diff
@@ -26,7 +26,7 @@
 namespace doris {
 
 std::string BetaRowset::segment_file_path(const std::string& dir, RowsetId rowset_id, int segment_id) {
```
```suggestion
std::string BetaRowset::segment_file_path(const std::string& dir, const RowsetId& rowset_id, const int segment_id) {
```

yiguolei replied:
Done

kangpinghuang commented on be/src/olap/rowset/rowset_meta.h:
```diff
@@ -71,12 +74,15 @@ class RowsetMeta {
 
-    void set_rowset_id(int64_t rowset_id) {
-        _rowset_meta_pb.set_rowset_id(rowset_id);
+    void set_rowset_id(RowsetId rowset_id) {
```
```suggestion
    void set_rowset_id(const RowsetId& rowset_id) {
```

kangpinghuang commented on be/src/olap/rowset/unique_rowset_id_generator.h:
```diff
@@ -0,0 +1,48 @@
+private:
+    SpinLock _lock;
+    UniqueId _backend_uid;
+    int64_t _version = 2; // modify it when create new version id generator
```
modify it to a const

yiguolei replied:
done

kangpinghuang commented on be/src/olap/rowset/unique_rowset_id_generator.cpp:
```diff
@@ -0,0 +1,50 @@
+    return _valid_rowset_ids.find(rowset_id) != _valid_rowset_ids.end();
+}
+
+void UniqueRowsetIdGenerator::release_id(RowsetId& rowset_id) {
```
```suggestion
void UniqueRowsetIdGenerator::release_id(const RowsetId& rowset_id) {
```

kangpinghuang commented on be/test/olap/rowset/unique_rowset_id_generator_test.cpp:
```diff
@@ -0,0 +1,101 @@
+}
+
+
+TEST_F(UniqueRowsetIdGeneratorTest, GenerateIdTest) {
```
add a case to test BE restart situation.

yiguolei replied:
Done

yiguolei committed:
Fix according to comment

imay commented on fe/src/main/java/org/apache/doris/task/SnapshotTask.java:
```diff
@@ -79,7 +79,7 @@ public TSnapshotRequest toThrift() {
         request.setVersion_hash(versionHash);
         request.setList_files(true);
-        request.setPreferred_snapshot_version(2);
+        request.setPreferred_snapshot_version(3);
```
Can make this a global variable? Its better not to write magic number

imay commented on be/test/olap/rowset/unique_rowset_id_generator_test.cpp:
```diff
@@ -0,0 +1,111 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
you should add this file to run-ut.sh

yiguolei replied:
Done

imay commented on be/src/olap/storage_engine.h:
```diff
@@ -333,6 +340,8 @@ class StorageEngine {
     std::unique_ptr<TabletManager> _tablet_manager;
     std::unique_ptr<TxnManager> _txn_manager;
 
+    RowsetIdGenerator* _rowset_id_generator;
```
should delete it in destructor or use std::unique_ptr

imay commented on be/src/olap/rowset/unique_rowset_id_generator.cpp:
```diff
@@ -0,0 +1,50 @@
+
+namespace doris {
+
+UniqueRowsetIdGenerator::UniqueRowsetIdGenerator(UniqueId backend_uid) :
```
```suggestion
UniqueRowsetIdGenerator::UniqueRowsetIdGenerator(const UniqueId& backend_uid) :
```

yiguolei replied:
Done

imay commented on be/src/olap/rowset/unique_rowset_id_generator.cpp:
```diff
@@ -0,0 +1,50 @@
+    _backend_uid(backend_uid), _inc_id(1) {
+}
+// generator a id according to data dir
+// rowsetid is not globally unique, it is dir level
```
I think rowset id is global unique

yiguolei replied:
Done

imay commented on be/src/olap/rowset/unique_rowset_id_generator.h:
```diff
@@ -0,0 +1,48 @@
+    // it saves the batch end id into meta env
+    OLAPStatus next_id(RowsetId* rowset_id) override; 
+
+    bool id_in_use(const RowsetId& rowset_id) override;
```
Do we really need this interface? If not we can make this class simpler.

imay commented on be/src/olap/rowset/segment_writer.cpp:
```diff
@@ -213,6 +213,12 @@ OLAPStatus SegmentWriter::finalize(uint32_t* segment_file_size) {
     std::string data_dir_string = data_dir_path.string();
     DataDir* data_dir = StorageEngine::instance()->get_store(data_dir_string);
+    data_dir->add_pending_ids(ROWSET_ID_PREFIX + _segment_group->rowset_id().to_string());
+    if (OLAP_SUCCESS != (res = file_handle.open_with_mode(
```
why we open this file two times?

imay commented on be/src/olap/rowset/segment_group.h:
```diff
@@ -48,11 +48,11 @@ namespace doris {
     friend class MemIndex;
 public:
-    SegmentGroup(int64_t tablet_id, int64_t rowset_id, const TabletSchema* tablet_schema,
+    SegmentGroup(int64_t tablet_id, RowsetId rowset_id, const TabletSchema* tablet_schema,
```
```suggestion
    SegmentGroup(int64_t tablet_id, const RowsetId& rowset_id, const TabletSchema* tablet_schema,
```

yiguolei replied:
Done

imay commented on be/src/olap/rowset/segment_group.h:
```diff
@@ -248,7 +248,7 @@ class SegmentGroup {
     int64_t get_tablet_id();
 
-    int64_t rowset_id() {
+    RowsetId rowset_id() {
```
```suggestion
    const RowsetId& rowset_id() const {
```

yiguolei replied:
Done

imay commented on be/src/olap/rowset/rowset_meta_manager.h:
```diff
@@ -31,20 +31,20 @@ namespace doris {
 class RowsetMetaManager {
 public:
-    static bool check_rowset_meta(OlapMeta* meta, TabletUid tablet_uid, int64_t rowset_id);
+    static bool check_rowset_meta(OlapMeta* meta, TabletUid tablet_uid, RowsetId rowset_id);
```
```suggestion
    static bool check_rowset_meta(OlapMeta* meta, TabletUid tablet_uid, const RowsetId& rowset_id);
```

yiguolei replied:
Done

morningman commented on be/src/service/doris_main.cpp:
```diff
@@ -153,6 +154,7 @@ int main(int argc, char** argv) {
     // options
     doris::EngineOptions options;
     options.store_paths = paths;
+    options.backend_uid = doris::UniqueId();
```
Is `backend_uid` act like a random seed that will be generated every time BE started?
If so, I think the name is confused that make people think of it as the ID of a BE?
Maybe `random_seed` is better?

morningman commented on be/src/olap/olap_define.h:
```diff
@@ -53,7 +53,7 @@ static constexpr uint32_t OLAP_COMPACTION_DEFAULT_CANDIDATE_SIZE = 10;
 static const uint16_t OLAP_STRING_MAX_LENGTH = 65535;
 
-static const int32_t PREFERRED_SNAPSHOT_VERSION = 2;
+static const int32_t PREFERRED_SNAPSHOT_VERSION = 3;
```
add some comment to explain the version 1 , 2, and 3?

yiguolei replied:
Done

imay commented on be/src/olap/data_dir.cpp:
```diff
@@ -582,7 +576,9 @@ OLAPStatus DataDir::_convert_old_tablet() {
             rowset_pb.SerializeToString(&meta_binary);
-            status = RowsetMetaManager::save(_meta, rowset_pb.tablet_uid(), rowset_pb.rowset_id() , meta_binary);
+            RowsetId rowset_id;
+            rowset_id.init(rowset_pb.rowset_id_v2());
```
if this is old version rwset_pb, this field will be 0. Is there no problem?

yiguolei replied:
No. You see this method is _convert_old_tablet. It is a method to convert old be before 0.10 to current be. During convert process, the rowset_id_v2 is set.

imay commented on be/src/olap/olap_common.h:
```diff
@@ -241,7 +246,83 @@ typedef std::set<uint32_t> UniqueIdSet;
+// 120 bit, backend random number generated by uuid
+// 64 bit, inc number from 0
+struct RowsetId {
+    int8_t version = 0;
```
you can use bit field

yiguolei committed:
Modify typo error according to review

yiguolei committed:
Add lock when check or release rowset id

yiguolei commented:
#1717

imay commented on be/src/olap/storage_engine.h:
```diff
@@ -196,6 +197,12 @@ class StorageEngine {
 
+    OLAPStatus next_rowset_id(RowsetId* rowset_id) { return _rowset_id_generator->next_id(rowset_id); };
+
+    bool rowset_id_in_use(RowsetId& rowset_id) { return _rowset_id_generator->id_in_use(rowset_id); };
```
```suggestion
    bool rowset_id_in_use(const RowsetId& rowset_id) { return _rowset_id_generator->id_in_use(rowset_id); };
```

imay commented on be/src/olap/rowset/segment_group.h:
```diff
@@ -262,12 +262,12 @@ class SegmentGroup {
     OLAPStatus copy_files_to(const std::string& dir);
 
-    OLAPStatus link_segments_to_path(const std::string& dest_path, int64_t rowset_id);
+    OLAPStatus link_segments_to_path(const std::string& dest_path, RowsetId rowset_id);
```
```suggestion
    OLAPStatus link_segments_to_path(const std::string& dest_path, const RowsetId& rowset_id);
```

imay commented on be/src/olap/rowset/segment_group.h:
```diff
@@ -248,7 +248,7 @@ class SegmentGroup {
     int64_t get_tablet_id();
 
-    int64_t rowset_id() {
+    const RowsetId& rowset_id() {
```
```suggestion
    const RowsetId& rowset_id() const {
```

yiguolei committed:
Using const & as parameter for rowsetid

imay reviewed (approved):
LGTM
